nextcord/http.py

```diff
@@ -124,25 +140,329 @@ def bucket(self) -> str:
+
+        if self._use_reset_timestamp:  # In reset timestamp mode.
+            if self.reset:
+                current_time = datetime.now(tz=timezone.utc)
+                reset_delta = (self.reset - current_time).total_seconds()
+                # Use estimated time between resets to guess when the next one will be.
+                # The non-zero comparison helps prevent issues when reset_diff isn't quite perfect yet.
+                if reset_delta < self._reset_ignore_threshold:
+                    seconds_until_reset = self._tracked_reset_time
+                else:  # Reset when the server resets.
+                    seconds_until_reset = reset_delta
+
+            else:
+                seconds_until_reset = self._tracked_reset_time
+
+        else:  # In reset_after seconds mode.
+            seconds_until_reset = self._tracked_reset_time
+
+        loop = asyncio.get_running_loop()
+        self._reset_remaining_task = loop.create_task(self.reset_remaining(seconds_until_reset))
+
+    async def reset_remaining(self, time: float) -> None:
+        """|coro|
+        Sleeps for the specified amount of time, then resets the remaining request count to the limit.
+
+        Parameters
+        ----------
+        time: :class:`float`
+            Amount of time to sleep until the request count is reset to the limit. ``time_offset`` is not added to
+            this number.
+        """
+        _log.debug("Bucket %s: Resetting after %s seconds.", self.bucket, time)
+        await asyncio.sleep(time)
+        self.remaining = self.limit
+        self._on_reset_event.set()
+        _log.debug("Bucket %s: Reset, allowing requests to continue.", self.bucket)
 
-    def __enter__(self) -> Self:
-        return self
+    @property
+    def migrating(self) -> Optional[str]:
+        """If not ``None``, this indicates what bucket acquiring requests should migrate to."""
+        return self._migrating
+
+    def migrate_to(self, bucket: str) -> None:
+        """Signals to acquiring requests, both present and future, that they need to migrate to a new bucket."""
+        self._migrating = bucket
+        self.remaining = self.limit
+        self._on_reset_event.set()
+        _log.debug(
+            "Bucket %s: Deprecating, acquiring requests will migrate to a new bucket.", bucket
+        )
 
-    def defer(self) -> None:
-        self._unlock = False
+    async def __aenter__(self) -> None:
+        await self.acquire()
 
-    def __exit__(
+    async def __aexit__(
         self,
-        exc_type: Optional[Type[BE]],
-        exc: Optional[BE],
-        traceback: Optional[TracebackType],
+        exc_type: Optional[Type[BaseException]],
+        exc_val: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
     ) -> None:
-        if self._unlock:
-            self.lock.release()
+        self.release()
+
+    @property
+    def locked(self) -> bool:
+        return self.remaining <= 0
+
+    async def acquire(self) -> bool:
```

Comment:
Requests which are in-progress don't count towards the rate limit. This may cause issues when the request is close to the reset time